core/authority_discovery/publisher/address_publisher.cpp

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "authority_discovery/publisher/address_publisher.hpp"
+
+#include "authority_discovery/protobuf/authority_discovery.v2.pb.h"
+
+#include "crypto/sha/sha256.hpp"
+
+#define _PB_SPAN(f) [&](gsl::span<const uint8_t> a) { (f)(a.data(), a.size()); }
+#define PB_SPAN_SET(a, b, c) _PB_SPAN((a).set_##b)(c)
+#define PB_SPAN_ADD(a, b, c) _PB_SPAN((a).add_##b)(c)
+
+template <typename T>
+std::vector<uint8_t> pbEncodeVec(const T &v) {
+  std::vector<uint8_t> r(v.ByteSizeLong());
+  v.SerializeToArray(r.data(), r.size());
+  return r;
+}
+
+namespace kagome::authority_discovery {
+  AddressPublisher::AddressPublisher(
+      std::shared_ptr<runtime::AuthorityDiscoveryApi> authority_discovery_api,
+      network::Roles roles,
+      std::shared_ptr<application::AppStateManager> app_state_manager,
+      std::shared_ptr<blockchain::BlockTree> block_tree,
+      std::shared_ptr<crypto::SessionKeys> keys,
+      const libp2p::crypto::KeyPair &libp2p_key,
+      const libp2p::crypto::marshaller::KeyMarshaller &key_marshaller,
+      std::shared_ptr<crypto::Ed25519Provider> crypto_provider,
+      std::shared_ptr<crypto::Sr25519Provider> crypto_provider2,
+      libp2p::Host &host,
+      std::shared_ptr<libp2p::protocol::kademlia::Kademlia> kademlia,
+      std::shared_ptr<libp2p::basic::Scheduler> scheduler)
+      : authority_discovery_api_(std::move(authority_discovery_api)),
+        roles_(roles),
+        block_tree_(std::move(block_tree)),
+        keys_(std::move(keys)),
+        crypto_provider_(std::move(crypto_provider)),
+        crypto_provider2_(std::move(crypto_provider2)),
+        host_(host),
+        kademlia_(std::move(kademlia)),
+        scheduler_(std::move(scheduler)),
+        log_{log::createLogger("AddressPublisher")} {
+    app_state_manager->atLaunch([=] { return start(); });
+    if (libp2p_key.privateKey.type == libp2p::crypto::Key::Type::Ed25519) {
+      libp2p_key_.emplace(crypto::Ed25519Keypair{
+          .secret_key =
+              crypto::Ed25519PrivateKey::fromSpan(libp2p_key.privateKey.data)
+                  .value(),
+          .public_key =
+              crypto::Ed25519PublicKey::fromSpan(libp2p_key.publicKey.data)
+                  .value(),
+      });
+      libp2p_key_pb_.emplace(
+          key_marshaller.marshal(libp2p_key.publicKey).value());
+    } else {
+      SL_WARN(log_, "Peer key is not ed25519");
```

Comment:
In case it has to be ed25519, would it not be better to replace constructor with a static build method and return an error?

Reply:
`libp2p::crypto::Keypair` comes from libp2p injector.